vision/diffusion_mnist/diffusion_mnist.jl

```diff
@@ -0,0 +1,385 @@
+        # https://pytorch.org/docs/stable/generated/torch.nn.ConvTranspose2d.html#convtranspose2d
+        #
+        # A fix is referenced here: https://github.com/FluxML/Flux.jl/issues/1319
+        # and here: https://github.com/FluxML/Flux.jl/issues/829
+        # But padding with `pad=SamePad()` or `pad=(0, 1, 0, 1)` yields (10, 10, 64, 3) 
+        # which is still incorrect.
+        # 
+        # Based off relationship 14 of https://arxiv.org/pdf/1603.07285.pdf:
+        # 𝘰' = 𝘴(𝘪' - 1) + 𝘢 + 𝘬 - 2𝘱
+        # Set 𝘰' = 12, 𝘴 = 2, 𝘪' = 5, 𝘬 = 3, 𝘱 = 0, then
+        # 12 = 2(5 - 1) + a + 3 - 2(0) = 11 + a => a = 1
+        # However, no Flux API that I could find exposes the argument a for the user...
+        #
+        # A (incorrect) hack to get the shapes to work is just to pad by (-1, 0, -1, 0):
+        # julia> randn(5, 5, 256, 3) |> ConvTranspose((3, 3), 256 => 64, stride=2, pad=(0, -1, 0, -1)) |> size
+        # (12, 12, 64, 3)
+        #
+        # Which is the correct shape, but seems suspicious (negative padding??).
+        # Why is passing a negative padding even allowed in the first place? 😕
+        ########################################################################
+        ConvTranspose((3, 3), channels[3] + channels[3] => channels[2], pad=(0, -1, 0, -1), stride=2, bias=false),
+        Dense(embed_dim, channels[2]),
+        GroupNorm(channels[2], 32),
+        ConvTranspose((3, 3), channels[2] + channels[2] => channels[1], pad=(0, -1, 0, -1), stride=2, bias=false),
+        Dense(embed_dim, channels[1]),
+        GroupNorm(channels[1], 32),
+        ConvTranspose((3, 3), channels[1] + channels[1] => 1, stride=1, bias=false),
+        # Scaling Factor
+        marginal_prob_std
+    )
+end
+
+@functor UNet
+
+"""
+Helper function that adds `dims` dimensions to the front of a `AbstractVecOrMat`.
+Similar in spirit to TensorFlow's `expand_dims` function.
+
+# References:
+https://www.tensorflow.org/api_docs/python/tf/expand_dims
+"""
+expand_dims(x::AbstractVecOrMat, dims::Int=2) = reshape(x, (ntuple(i -> 1, dims)..., size(x)...))
+
+"""
+Helper function that reverses the order of dimensions.
+"""
+reverse_dims(x) = permutedims(x, reverse(ntuple(i -> i, length(size(x)))))
+
+"""
+Makes the UNet struct callable and shows an example of a "Functional" API for modeling in Flux. \n
+
+Notes:
+    ```julia
+    @assert size(a) == (28, 28, 1, 32, 32)
+    @assert size(b) == (32, 32)
+    o = reverse_dims(a) .+ reverse_dims(b)' |> reverse_dims
+    size(o)  # (28, 28, 1, 32, 32)
+    ```
+    Is used to broadcast an operation from b to a 
+    without the cost of a copy (using expand_dims for example)
+    which is not allowed (mutating vector) by Zygote in 
+    a custom model.
+"""
+function (unet::UNet)(x, t)
+    # Embedding
+    embed = unet.gaussfourierproj(t)
+    embed = swish.(unet.linear(embed))
+    # Encoder
+    h1 = unet.conv1(x)
+    h1 = reverse_dims(h1) .+ unet.dense1(embed)' |> reverse_dims
```

Comment:
No, this will work fine! I will go ahead and get rid of the `reverse_dims` sillyness, as this is unneeded.

Full story, I originally had:
``` julia
h1 .+= expand_dims(unet.dense1(embed), 2)
``` 
and getting
```julia
ERROR: Mutating arrays is not supported -- called copyto!(::Array{Float32, 4}, _...)
```
which I **incorrectly** deduced `expand_dims` being a copy and the source of the issue.

I thought it was odd reshape was doing a copy... now I understand its the `.+=` operator 🤦‍♂️

Reply:
`h_x = reverse_dims(h_x) .+ unet.dense_x(embed)' |> reverse_dims` (throughout entire network):
```julia
julia>     @time [unet_test(x_test, t_test) for _ in 1:500];
 64.377195 seconds (540.61 k allocations: 41.847 GiB, 0.88% gc time, 0.05% compilation time)
```

`h_x = h_x .+ expand_dims(unet.dense_x(embed), 2)` (throughout entire network):
```julia
julia>     @time [unet_test(x_test, t_test) for i in 1:500];
 64.515426 seconds (529.79 k allocations: 33.513 GiB, 0.27% gc time, 0.04% compilation time)
```
So your suspicion was correct, it is less efficient memory-wise **and** way less pretty style-wise.

Reply:
FYI that https://github.com/JuliaCI/BenchmarkTools.jl will make your life much easier for this kind of timing ;)

Reply:
For benchmarks taking >3s, its usually the case that `@time` is sufficient